feat(db): Add "naive" PostgreSQL implementation

silvestrst-crypto opened the pull request:
### Description

Copy/paste of the SQLite implementation, and necessary changes to SQL syntax and data types.

The rationale for choosing Postgres was the fact that `rusqlite` crate was based on `postgres` crate, thus sharing similarities in the API. They have, however, diverged significantly, so this turned out to be not as much of advantage as was expected initially.

Resources:
https://docs.rs/postgres/0.19.4/postgres/index.html
https://docs.rs/rusqlite/0.27.0/rusqlite/index.html

**Notable differences:**
  - PostgreSQL does not have u64 type, and u32 is used for OIDs, so this implementation stores u64 as i64, and u32 as i32. On retrieval, these are converted back to uXX.
  - PostgreSQL does not have `INSERT OR REPLACE`, instead "upsert" syntax is used, which is `INSERT ON CONFLICT DO SET ...`.
  - PostgreSQL uses `$` to specify variables in the queries, instead of SQLite `:varname` syntax.
  - PostgreSQL types differ from SQLite (INT = i32, BIGINT = i64, ...).

**Design decisions:**
  - u64 and u32 types are converted to i64 and i32. It is assumed that no loss occurs by casting these types to and back.
  - Every test creates a new random database.

**Known shortcomings:**
  - At present, test databases are not deleted after tests have finished, and are removed manually via `psql` if required.


**PLEASE NOTE:
I AM NOT INTENDING TO MERGE THIS CHANGE INTO `[bbv-master-v0.26.0-eac739d](https://github.com/monacohq/bdk/tree/bbv-master-v0.26.0-eac739d)` THIS PR IS FOR REVIEW ONLY, AND THEN WILL LIVE IN THIS SEPARATE BRANCH.**

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [ ] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [ ] I've added docs for the new feature
* [ ] I've updated `CHANGELOG.md`

silvestrst-crypto committed:
feat(db): Add "naive" PostgreSQL implementation

Copy/paste of the SQLite implementation, and necessary changes
to SQL syntax and data types.

The rationale for choosing Postgres was the fact that `rusqlite`
crate was based on `postgres` crate, thus sharing similarities
in the API. They have, however, diverged significantly, so this
turned out to be not as much of advantage as was expected
initially.

Notable differences:
  - PostgreSQL does not have u64 type, and u32 is used for OIDs,
    so this implementation stores u64 as i64, and u32 as i32.
    On retrieval, these are converted back to uXX.
  - PostgreSQL does not have `INSERT OR REPLACE`, instead
    "upsert" syntax is used, which is `INSERT ON CONFLICT DO SET ...`.
  - PostgreSQL uses `$` to specify variables in the queries, instead
    of SQLite `:varname` syntax.
  - PostgreSQL types differ from SQLite (INT = i32, BIGINT = i64, ...).

Design decisions:
  - u64 and u32 types are converted to i64 and i32. It is assumed
    that no loss occurs by casting these types to and back.
  - Every test creates a new random database.

Known shortcomings:
  - At present, test databases are not deleted after tests have
    finished, and are removed manually via `psql` if required.

jimmyyip-crypto commented on src/database/postgres_db.rs:
```diff
@@ -0,0 +1,1253 @@
+    }
+
+    fn set_raw_tx(&mut self, transaction: &Transaction) -> Result<(), Error> {
+        match self.select_transaction_by_txid(&transaction.txid())? {
```
seems most of the `set_xxx` are using similiar pattern. the update is not atomic and there are no unique indexes on some tables.
we may try to synchronize multiple instances at the same time to see if there are any issues.

silvestrst-crypto replied:
Let me double check, I think that all operations are part of batch:
```
BEGIN TRANSACTION;
...
...
set_xxx
...
END TRANSACTION;
```

silvestrst-crypto replied:
Wallet for example uses `BatchDatabase`.

```
impl<D> Wallet<D>
where
    D: BatchDatabase,
{
...
```
```
/// Trait for a database that supports batch operations
///
/// This trait defines the methods to start and apply a batch of operations.
pub trait BatchDatabase: Database {
    /// Container for the operations
    type Batch: BatchOperations;

    /// Create a new batch container
    fn begin_batch(&self) -> Self::Batch;
    /// Consume and apply a batch of operations
    fn commit_batch(&mut self, batch: Self::Batch) -> Result<(), Error>;
}
```

But...

I don't think that higher level framework is using `batch.begin and batch.commit` :cold_sweat:

silvestrst-crypto replied:
Actually, It does, at least for syncing, but it is not on the same level, so it's hard to trace...

So it's like this:
Wallet(wallet_sync -> wallet_setup) -> ElectrumBlockchain/EsploraBlockchain(wallet_setup) -> script_sync.rs(satisfy->into_db_update) -> DB(begin_batch) ... ElectrumBlockchain/EsploraBlockchain ... -> DB(commit_batch)

silvestrst-crypto committed:
feat(database): Add PostgreSQL to AnyDatabase

silvestrst-crypto committed:
feat(database): Add TLS mode to PostgreSQL database backend

Please not that current implementation is susceptible to MITM attacks.
Although the data in transit is encrypted, the client doesn't verify
the certificate.

See `OpenSSL::SSL::VERIFY_NONE` for more information.

silvestrst-crypto commented:
Should implement mongodb or postgres for BDK v1.0.
